### DISCUSSION TITLE
Support HTTP/2 Server Push

### DISCUSSION THREAD

**Discussion by karpetrosyan** [encode/httpcore#828] [Ideas]
HTTP/2 Server push is a feature of HTTP/2 and HTTP/3 that **allows the server to send responses before the client requests them**. Because the response can be taken from the cache, it can significantly reduce response time. 
It is especially **useful when the server predicts which resource the client will request the next time**.

You can see the [wiki about the http/2 server push](https://en.wikipedia.org/wiki/HTTP/2_Server_Push) for more information.

`HTTP Core` does support the HTTP2 protocol, but it cannot support server push mechanisms because it also requires RFC 9111 implementation and some storage logic, which is not what we want to see in the `HTTP Core` library, so it **should just provide some API to do that**.

Here is an example of how it can be implemented.

```python
import httpcore

class ConcreteServerPush(httpcore.BaseServerPush):
   def store_response(self, response: httpx.Response):
       # some RFC 9111 and filesystem code

client = httpcore.ConnectionPool()
client.request("GET", "https://example.com", extensions={"server_push": ConcreteServerPush()}
```

The obvious question here is: Why should we support this?

There are a few reasons why we should support this.

- Very easy to implement. We are just providing an API for third-party libraries.
- All the modern browsers and HTTP servers do support HTTP/2 server push.
- Very fast response times (responses can literally be taken from the filesystem)

**Comment by lovelydinosaur** [comment]
Ah yes, interesting suggestion.

Although... Chrome have now removed HTTP/2 Server Push support, essentially in favour of the "103 Early Hints" status code.
Good background reading https://blog.cloudflare.com/the-state-of-http-in-2022/
